### PR TITLE
DEX-677 Wrong order cancel

### PR DESCRIPTION
**Pull request by vsuharnikov**
Fee: an object with functions to calculate fees;
OrderExecuted: counterExecutedFee and submittedRemainingFee now are static and not calculated;
ExchangeTransactionCreator now uses calculated fees from the OrderExecuted event;

CancelOrderTestSuite:
* A new integration test to check this bug;
* Test stabilization;

ExchangeTransactionCreatorSpecification: some tests migrated to FeeSpecification and OrderBookActorSpecification;

### CODE REVIEW

File: dex/src/test/scala/com/wavesplatform/dex/market/OrderBookActorSpecification.scala [hunk 1]
```diff
@@ -390,6 +394,31 @@ class OrderBookActorSpecification
...
+          Map(
+            0L -> DynamicSettings(0.001.waves, 0.003.waves),
+            1L -> DynamicSettings(0.001.waves, 0.005.waves)
+          ))) _
+
+      obcTestWithPrepare(prepare = (_, _) => (), makerTakerFeeAtOffset = makerTakerFeeAtOffset) { (pair, orderBook, tp) =>
+        tp.expectMsg(OrderBookRecovered(pair, None))
+        val now = System.currentTimeMillis()
+
+        val maker1 = sell(wavesUsdPair, 10.waves, 3.00, matcherFee = 0.003.waves.some, ts = now.some) // was placed when order-fee was DynamicSettings(0.003.waves, 0.003.waves)
```
Comment (DenisSerebryansky) [inline dex/src/test/scala/com/wavesplatform/dex/market/OrderBookActorSpecification.scala:409]: then order fee settings was `DynamicSettings(0.001.waves, 0.003.waves)` actually

File: dex-it/src/test/scala/com/wavesplatform/it/sync/CancelOrderTestSuite.scala [hunk 2]
```diff
@@ -242,8 +201,36 @@ class CancelOrderTestSuite extends MatcherSuiteBase {
...
+      dex1.api.orderStatus(counterOrder).status shouldBe OrderStatus.PartiallyFilled
+      dex1.api.orderStatus(submittedOrder).status shouldBe OrderStatus.Filled
+    }
+
+    "wrong cancel when match on all coins" in {
+      val accounts = (1 to 30).map(i => KeyPair(s"auto-cancel-$i".getBytes(StandardCharsets.UTF_8)))
+      knownAccounts = knownAccounts ++ accounts
+
       val oneOrderAmount = 10000
       val orderPrice     = 3000000000000L
```
[the remark this answers is not in the document]
Comment (vsuharnikov) [reply dex-it/src/test/scala/com/wavesplatform/it/sync/CancelOrderTestSuite.scala:235]: + :)

File: dex/src/main/scala/com/wavesplatform/dex/model/MatcherModel.scala [hunk 3]
```diff
@@ -77,11 +77,15 @@ sealed trait AcceptedOrder {
...
   }
+
+  def isFirstMatch: Boolean = amount == order.amount
 }
 
 object AcceptedOrder {
 
   def partialFee(matcherFee: Long, totalAmount: Long, partialAmount: Long): Long = {
+    if (partialAmount > totalAmount)
+      throw new IllegalArgumentException(s"partialAmount: $partialAmount should be less or equal to totalAmount: $totalAmount")
```
[the remark this answers is not in the document]
Comment (vsuharnikov) [reply dex/src/main/scala/com/wavesplatform/dex/model/MatcherModel.scala:88]: Yeah, this is a pre-condition. It helped to find a bug in percent fee calculation

File: dex/src/main/scala/com/wavesplatform/dex/model/MatcherModel.scala [hunk 4]
```diff
@@ -308,23 +312,22 @@ object Events {
...
-    *  In case of dynamic fee settings the following params can be different from the appropriate `acceptedOrder.order.matcherFee`
-    * @param maxSubmittedFee limited by base-taker-fee
-    * @param maxCounterFee limited by base-maker-fee
+    * In case of dynamic fee settings the following params can be different from the appropriate `acceptedOrder.order.matcherFee`
     */
-  case class OrderExecuted(submitted: AcceptedOrder, counter: LimitOrder, timestamp: Long, maxSubmittedFee: Long, maxCounterFee: Long) extends Event {
+  case class OrderExecuted(submitted: AcceptedOrder, counter: LimitOrder, timestamp: Price, counterExecutedFee: Price, submittedExecutedFee: Price)
+      extends Event {
 
+    def executedPrice: Long                   = counter.price
```
[the remark this answers is not in the document]
Comment (vsuharnikov) [reply dex/src/main/scala/com/wavesplatform/dex/model/MatcherModel.scala:320]: It doesn't matter. I prefer `def` for such situations when a RHS is fixed.

File: dex/src/main/scala/com/wavesplatform/dex/model/MatcherModel.scala [hunk 5]
```diff
@@ -308,23 +312,22 @@ object Events {
   sealed trait Event
 
   /**
-    *  In case of dynamic fee settings the following params can be different from the appropriate `acceptedOrder.order.matcherFee`
-    * @param maxSubmittedFee limited by base-taker-fee
-    * @param maxCounterFee limited by base-maker-fee
+    * In case of dynamic fee settings the following params can be different from the appropriate `acceptedOrder.order.matcherFee`
     */
-  case class OrderExecuted(submitted: AcceptedOrder, counter: LimitOrder, timestamp: Long, maxSubmittedFee: Long, maxCounterFee: Long) extends Event {
+  case class OrderExecuted(submitted: AcceptedOrder, counter: LimitOrder, timestamp: Price, counterExecutedFee: Price, submittedExecutedFee: Price)
```
[the remark this answers is not in the document]
Comment (vsuharnikov) [reply dex/src/main/scala/com/wavesplatform/dex/model/MatcherModel.scala:317]: Haha, IDEA changed it to the Price during fields reordering! :D I'll change it

File: dex/src/main/scala/com/wavesplatform/dex/model/OrderValidator.scala [hunk 6]
```diff
@@ -179,7 +179,7 @@ object OrderValidator extends ScorexLogging {
 
       lazy val exchangeTx: Result[ExchangeTransaction] = {
         val fakeOrder: Order  = order.updateType(order.orderType.opposite)
-        val oe: OrderExecuted = OrderExecuted(LimitOrder(fakeOrder), LimitOrder(order), time.correctedTime(), fakeOrder.matcherFee, order.matcherFee)
+        val oe: OrderExecuted = OrderExecuted(LimitOrder(fakeOrder), LimitOrder(order), time.correctedTime(), order.matcherFee, fakeOrder.matcherFee)
```
[the remark this answers is not in the document]
Comment (vsuharnikov) [reply dex/src/main/scala/com/wavesplatform/dex/model/OrderValidator.scala:182]: Ok, it is shorter :)

File: dex/src/test/scala/com/wavesplatform/dex/MatcherSpecBase.scala [hunk 7]
```diff
@@ -416,4 +417,29 @@ trait MatcherSpecBase extends NTPTime with DiffMatcherWithImplicits with DoubleO
 
     Order.sign(correctedOrder, sender)
   }
+
+  def mkOrderExecutedRaw(submitted: Order, counter: Order): OrderExecuted =
```
[the remark this answers is not in the document]
Comment (vsuharnikov) [reply dex/src/test/scala/com/wavesplatform/dex/MatcherSpecBase.scala:421]: Sure

File: dex/src/main/scala/com/wavesplatform/dex/model/MatcherModel.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex/src/main/scala/com/wavesplatform/dex/model/MatcherModel.scala:320]: image_2020-04-02_10-12-55

[the remark this answers is not in the document]
Comment (vsuharnikov) [reply dex/src/main/scala/com/wavesplatform/dex/model/MatcherModel.scala:320]: right-hand side of the expression :D

**Pull request merged**